### PR TITLE
[tex] backport backup part from cloud6

### PR DESCRIPTION
**Pull request by MaximilianMeister**
take only the necessary parts to make a backup

the related PR is https://github.com/crowbar/crowbar-core/pull/144

### CODE REVIEW

**Comment by MaximilianMeister** [comment]
the idea is that the cloud5 user can make a backup which can be used to be imported into a new cloud6

**Comment by aplanas** [comment]
Isn't a cherry-pick a better option?

**Comment by MaximilianMeister** [comment]
@aplanas not here, we take only the necessary parts to make a backup. and the parent commits have more than what we need here.
and on top of that, the other PR is not merged yet and the commit id's are likely to change...

**Comment by rsalevsky** [comment]
@aplanas And we have different repos so cherry-pick wouldn't work. ;)

File: crowbar_framework/Gemfile [hunk 1]
```diff
@@ -35,6 +35,8 @@ gem "simple-navigation", "~> 3.12"
 gem "simple_navigation_renderers", "~> 1.0"
 gem "sqlite3", "~> 1.3"
 gem "syslogger", "~> 1.6"
+gem "yaml_db", "~> 0.3.0"
```
Comment (aplanas) [inline crowbar_framework/Gemfile]: Looks like that gaiting is complaining about this: "`resolve': Could not find gem 'yaml_db (~> 0.3.0) ruby'`"

**Comment by tboerger** [comment]
:+1: but gating fails

**Comment by MaximilianMeister** [comment]
@tboerger that's because @rsalevsky can get the missing dependencies as an update into cloud5 after this is merged IIUC.

**Comment by MaximilianMeister** [comment]
cloud5backup

**Comment by rhafer** [comment]
After applying and restarting crowbar I get this:

```
F, [2015-12-17T14:18:59.453306 #2883] FATAL -- : 
ActionView::Template::Error (translation missing: en.nav.utils.backup):
    11:       = link_to page_title, root_path, :class => "navbar-brand"
    12: 
    13:     .collapse.navbar-collapse.navbar-right
    14:       = render_navigation :expand_all => true, :skip_if_empty => true
```

So it seem some change to `crowbar_framework/config/locales/crowbar/en.yml` is missing from the PR.

**Comment by rhafer** [comment]
Next failure:

```
Errno::ENOENT (No such file or directory @ dir_initialize - /var/lib/crowbar/backup):
  app/models/backup.rb:73:in `all'
  app/controllers/backup_controller.rb:26:in `index'
```

Though I guess it would also be fine if that directory is just part of the rpm package.

**Comment by rhafer** [comment]
After fix those two simple things I get a backup created. But even though the file is called `bla-bla-bla.tar.gz` it's not gzip compressed it a simple (uncompressed) tar file.

File: crowbar_framework/app/models/backup.rb [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  def create
+    saved = false
+    dir = Dir.mktmpdir
+
+    Crowbar::Backup::Export.new(dir).export
+    Dir.chdir(dir) do
+      ar = Archive::Compress.new(
+        "#{Backup.image_dir}/#{name}-#{created_at}.tar.gz",
+        type: :tar,
+        compression: :gzip2
```
Comment (rhafer) [inline crowbar_framework/app/models/backup.rb]: I guess that's either `:bzip2` or `:gzip`, would explain my previous comment.

Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: good catch. thanks I haven't noticed yet

**Comment by MaximilianMeister** [comment]
> So it seem some change to crowbar_framework/config/locales/crowbar/en.yml is missing from the PR.

i think this is in crowbar.yml and after a `Rake assets:precompile` the error may be gone

**Comment by MaximilianMeister** [comment]
> Though I guess it would also be fine if that directory is just part of the rpm package.

~~there is already a submit request from @rsalevsky IIRC~~ merged

**Comment by rhafer** [comment]
> So it seem some change to crowbar_framework/config/locales/crowbar/en.yml is missing from the PR.
> 
> i think this is in crowbar.yml and after a Rake assets:precompile the error may be gone

No, that didn't help here. I really needed to add that key to the en.yml in locales

**Comment by MaximilianMeister** [comment]
> No, that didn't help here. I really needed to add that key to the en.yml in locales

ok I will add it then for now

**Comment by MaximilianMeister** [comment]
@rhafer I hope we have fixed all issues for now, lets wait for the gate

File: crowbar.yml [hunk 3]
```diff
@@ -73,6 +73,9 @@ nav:
     logs:
       order: 10
       route: 'utils_path'
+    backup:
+      order: 20
+      route: 'backup_path'
```
Comment (rhafer) [inline crowbar.yml]: I think we should not offer this feature in the utils menu in Tex. It's meant for creating dump that can only be used for upgrading to Master. The dump cannot be restored on Cloud 5, right? If that's true, offering it in the "Utils" menu is creating wrong expectations I think.
The backup for upgrade should be triggerd with the upgrade wizard (whenever that is implemented)

Comment (MaximilianMeister) [inline crowbar.yml]: makes sense somehow. but I would rather not drop the link completely. Maybe we can just add a flash message which says sth like "THis Backup Page is meant only to assist the Upgrade to Cloud 6" ?

Comment (tboerger) [inline crowbar.yml]: Until we get the wizard I would prefer to keep this link and add a message at the top that clearly states the use case of this page.

File: crowbar_framework/lib/crowbar/backup/base.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            [
+              "configs/crowbar.json",
+              "/etc/crowbar/crowbar.json"
+            ],
+            [
+              "configs/hosts",
+              "/etc/hosts"
+            ],
+            [
+              "configs/HOSTNAME",
```
Comment (rsalevsky) [inline crowbar_framework/lib/crowbar/backup/base.rb]: There is a difference between SLE11SP3 and SLE12SP1, so we should save the file to `configs/hostname`.

File: crowbar_framework/lib/crowbar/backup/export.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+                file.write("nameserver #{forwarder}")
+              end
+            end
+          else
+            system("sudo", "-i", "cp", "-r", source, data_dir.join(destination).to_s)
+            system("sudo", "-i", "chown", "-R", "crowbar:crowbar", source, data_dir.join(destination).to_s)
+          end
+        end
+
+        data_dir.join("version").open("w") do |file|
```
Comment (rsalevsky) [inline crowbar_framework/lib/crowbar/backup/export.rb]: This part can be dropped, we have now the meta.

**Comment by rhafer** [comment]
I think it makes sense to squash all of this into a single commit. All of the subsequent commits are basically fixups of the first commit. After that I think this code is fine. (We need to do something about the ci though)

File: crowbar.yml (hunk 3 above)
Comment (vuntz) [inline crowbar.yml]: I've added a task to the upgrade UI card to drop this.

File: crowbar_framework/lib/crowbar/backup.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+#
+
+module Crowbar
+  class Backup
+    autoload :Base,
+      File.expand_path("../backup/base", __FILE__)
+
+    autoload :Export,
+      File.expand_path("../backup/export", __FILE__)
+  end
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup.rb:25]: Stupid question: why is this needed?

File: crowbar_framework/lib/crowbar/backup/base.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+              "configs/hosts",
+              "/etc/hosts"
+            ],
+            [
+              "configs/hostname",
+              "/etc/HOSTNAME"
+            ],
+            [
+              "configs/network.json",
+              "/etc/crowbar/network.json"
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/base.rb]: See https://github.com/crowbar/crowbar-core/pull/144/files#r48938161

File: crowbar_framework/lib/crowbar/backup/export.rb [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+          "role",
+          "::Chef::Role"
+        )
+      end
+
+      def databags
+        logger.debug "Backing up databags"
+
+        data_dir = workdir.join("knife", "databags")
+        data_dir.mkpath unless data_dir.directory?
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/export.rb]: See https://github.com/crowbar/crowbar-core/pull/144/files#r48938462

File: crowbar_framework/app/views/backup/index.html.haml [hunk 9]
```diff
@@ -0,0 +1,52 @@
+.row
```
Comment (vuntz) [inline crowbar_framework/app/views/backup/index.html.haml]: FWIW, I don't think we will need the full UI (as there's no restore for Cloud 5). It's useful for now, but once we get the upgrade UI, I think we can drop that.

File: crowbar_framework/app/controllers/backup_controller.rb [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  def backup
+    @backup = Backup.new(
+      name: params[:filename]
+    )
+    respond_to do |format|
+      if !@backup.nil? && @backup.save
+        format.json { head :ok }
+        format.html { redirect_to backup_path }
+      else
+        msg = I18n.t(".invalid_filename", scope: "backup.index")
```
Comment (vuntz) [inline crowbar_framework/app/controllers/backup_controller.rb]: See https://github.com/crowbar/crowbar-core/pull/144/files#r48939967

File: crowbar_framework/app/controllers/backup_controller.rb [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+  # data-confirm method delete
+  # /utils/backup/delete 	DELETE 	Delete a backup
+  def delete
+    @backup = Backup.new(
+      name: params[:name],
+      created_at: params[:created_at]
+    )
+
+    respond_to do |format|
+      if @backup.valid?
```
Comment (vuntz) [inline crowbar_framework/app/controllers/backup_controller.rb:82]: https://github.com/crowbar/crowbar-core/pull/144/files#r48940093

File: crowbar_framework/app/models/backup.rb [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+    def image_dir
+      if Rails.env.production?
+        Pathname.new("/var/lib/crowbar/backup")
+      else
+        Rails.root.join("storage")
+      end
+    end
+
+    def filename_time(filename)
+      filename.split(/([\w-]+)-([0-9]{8}-[0-9]{6})/).reject(&:empty?)
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: This is not the same regexp as `filename_has_characters` for the filename.

File: crowbar_framework/app/models/backup.rb [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+    def all
+      list = []
+
+      backup_files = image_dir.children.select do |c|
+        c.file? && c.to_s =~ /gz$/
+      end
+
+      backup_files.each do |backup_file|
+        name, created_at = filename_time(backup_file.basename.to_s)
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Shouldn't this be `name, created_at, _ext =` ?

File: crowbar_framework/app/models/backup.rb [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+  attr_accessor :name, :created_at, :filename, :path
+
+  validates_presence_of :name, :created_at
+  validates :created_at, exclusion: { in: [nil, ""] }
+  validate :filename, :filename_has_format
+  validate :filename, :filename_has_characters
+
+  def initialize(options)
+    @name = options.fetch :name, nil
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Shouldn't there be a default instead of `nil` (like `crowbar-backup`)?

File: crowbar_framework/app/models/backup.rb [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+  end
+
+  def create
+    saved = false
+    dir = Dir.mktmpdir
+
+    Crowbar::Backup::Export.new(dir).export
+    Dir.chdir(dir) do
+      system(
+        "tar czf #{Backup.image_dir}/#{name}-#{created_at}.tar.gz *"
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Why not use `@path` here?

File: crowbar_framework/app/models/backup.rb [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    def all
+      list = []
+
+      backup_files = image_dir.children.select do |c|
+        c.file? && c.to_s =~ /gz$/
+      end
+
+      backup_files.each do |backup_file|
+        name, created_at = filename_time(backup_file.basename.to_s)
+        image = new(
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb]: Would name this variable backup, not image.

**Comment by rhafer** [comment]
~~>This part can be dropped, we have now the meta. (https://github.com/crowbar/barclamp-crowbar/pull/1359#discussion_r48841846)~~

~~Hm, now restore on Cloud complains about missing version file.~~

I guess this is a bug in the restore part on master, I'll move my comment there.

File: crowbar_framework/lib/crowbar/backup/export.rb [hunk 17]
```diff
@@ -0,0 +1,173 @@
...
+
+        klass.constantize.list.each do |name, url|
+          logger.debug "Backing up #{component} #{name}"
+
+          record = klass.constantize.load(
+            name
+          ).to_hash.merge(
+            json_class: klass.constantize.to_s
+          )
```
Comment (rhafer) [inline crowbar_framework/lib/crowbar/backup/export.rb]: See: https://github.com/crowbar/crowbar-core/pull/144/files#r49166530

File: crowbar_framework/app/models/backup.rb (hunk 12 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: I have dropped the validations, not needed anymore

File: crowbar_framework/app/models/backup.rb (hunk 13 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: I have just not taken care about caching the extension because we don't really need it

File: crowbar_framework/app/controllers/backup_controller.rb [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+#
+
+class BackupController < ApplicationController
+  #
+  # Backup
+  #
+  # Provides the restful api call for
+  # /utils/backup   POST   Trigger a backup
+  def backup
+    @backup = Backup.new
```
Comment (rhafer) [inline crowbar_framework/app/controllers/backup_controller.rb:24]: Don't you need to pass parameters here? The model is expecting some.

Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/backup_controller.rb:24]: not anymore, I have added default values in `initialize`

**Comment by MaximilianMeister** [comment]
I have further dropped some unused code

File: crowbar_framework/app/controllers/backup_controller.rb (hunk 18 above)
Comment (rhafer) [inline crowbar_framework/app/controllers/backup_controller.rb:24]: Alternatively you could of course define defaults in the Model's constructor I guess.

File: crowbar_framework/app/models/backup.rb [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+
+class Backup
+  include ActiveModel::Model
+
+  attr_accessor :name, :created_at, :filename, :path
+
+  validates_presence_of :name, :created_at
+  validates :created_at, exclusion: { in: [nil, ""] }
+
+  def initialize(options)
```
Comment (rhafer) [inline crowbar_framework/app/models/backup.rb]: Ok, I guess you're missing a `:{}` here. Without that I get: ArgumentError (wrong number of arguments (0 for 1))

File: crowbar_framework/app/controllers/backup_controller.rb (hunk 18 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/controllers/backup_controller.rb:24]: > Alternatively you could of course define defaults in the Model's constructor I guess.

yes that's exactly how i did it https://github.com/crowbar/barclamp-crowbar/pull/1359/files#diff-408ac208d889893305de65518aa046a2R27
the second parameter to `fetch` is the fallback default value

File: crowbar_framework/app/models/backup.rb (hunk 19 above)
Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: hmm strange it worked for me, but i guess you meant adding `options = {}`. okay will do

Comment (MaximilianMeister) [inline crowbar_framework/app/models/backup.rb]: oh i see in `as_json` we have it like that as well. new Ruby hash syntax. you're right :)

**Comment by rhafer** [comment]
> oh i see in as_json we have it like that as well. new Ruby hash syntax. you're right :)

Actually I was not quite right. Now the download method is broken.

**Comment by rhafer** [comment]
I guess you'd either need to switch to ruby 2 style keyword arguments. Or use the old syntax `options = {}`

**Comment by MaximilianMeister** [comment]
> Actually I was not quite right. Now the download method is broken.

thanks for the hint. fixed

**Comment by MaximilianMeister** [comment]
> I guess you'd either need to switch to ruby 2 style keyword arguments. Or use the old syntax options = {}

it was actually the missing `exist?` method for the backup object, i have just added `@backup.path.exist?`

**Comment by rhafer** [comment]
> it was actually the missing exist? method for the backup object, i have just added @backup.path.exist?

I still get 

```
ArgumentError (unknown keywords: name, created_at):
  app/controllers/backup_controller.rb:46:in `new'
  app/controllers/backup_controller.rb:46:in `download'
```

When calling `curl -u crowbar:crowbar http://localhost:3000/utils/backup/download/crowbar_upgrade_data/20160113-162124 -v --digest`

Am I using the wrong command? It goes away when I change to "options={}" in the model's constructor.

**Comment by MaximilianMeister** [comment]
> Am I using the wrong command? It goes away when I change to "options={}" in the model's constructor.

I was a bit confused as well, `options = {}` is correct here

File: crowbar_framework/lib/crowbar/backup/base.rb [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+              "root/.gnupg",
+              "/root/.gnupg"
+            ],
+            [
+              "root/.ssh",
+              "/root/.ssh"
+            ],
+            [
+              "data",
+              "/var/lib/crowbar"
```
Comment (rhafer) [inline crowbar_framework/lib/crowbar/backup/base.rb:44]: This leads to all previous backups being included in the current backup. We should avoid that.

File: crowbar_framework/app/controllers/backup_controller.rb [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+          send_file(
+            @backup.path,
+            filename: "#{params[:name]}-#{params[:created_at]}.tar.gz"
+          )
+        end
+      else
+        msg = I18n.t(".missing_backup", scope: "backup")
+        format.html do
+          flash[:alert] = msg
+          redirect_to backup_path
```
Comment (rhafer) [inline crowbar_framework/app/controllers/backup_controller.rb:62]: `NameError (undefined local variable or method `backup_path'`

I think this is also true in a number of other places where the backup_path helper is used.

**Comment by rhafer** [comment]
There are still a few issues with the code, especially in the error handling it seems. And depending on the design of the Upgrade wizard it might still require some changes.

But to me it would be ok to merge in the current state so that people can start playing with it more.

So +1 from me and lets address the remaining bit in subsequent pull requests.

File: crowbar_framework/config/locales/crowbar/en.yml [hunk 22]
```diff
@@ -536,14 +536,23 @@ en:
       two_words_connector: ' and '
       last_word_connector: ' and '
 
+  backup:
+    backup: 'Backup'
+    crowbar: 'Crowbar'
+    openstack: 'OpenStack'
+    crowbar_openstack: 'Crowbar & OpenStack'
```
Comment (vuntz) [inline crowbar_framework/config/locales/crowbar/en.yml:543]: The strings above are not used, it seems?

File: crowbar_framework/app/models/backup.rb [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+
+    Crowbar::Backup::Export.new(dir).export
+    Dir.chdir(dir) do
+      system(
+        "tar czf #{path} *"
+      )
+      saved = true
+    end
+    FileUtils.rm_rf(dir)
+    saved
```
Comment (vuntz) [inline crowbar_framework/app/models/backup.rb:74]: Why do we need the `saved` variable?

**Comment by vuntz** [comment]
> But to me it would be ok to merge in the current state so that people can start playing with it more.

Same here, I'm okay merging if there's a followup PR to address the comments.

File: crowbar_framework/lib/crowbar/backup/export.rb [hunk 24]
```diff
@@ -0,0 +1,177 @@
...
+            data_dir.join(destination).open("w") do |file|
+              forwarders.each do |forwarder|
+                file.write("nameserver #{forwarder}")
+              end
+            end
+          else
+            # copy files with higher permissions
+            dest = data_dir.join(destination).to_s
+            system("sudo", "cp", "-a", source, dest)
+            system("sudo", "chown", "-R", "crowbar:crowbar", dest)
```
Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/export.rb:110]: (just as a reminder from the other PR: this bit will really cause issues later on, we need to keep permissions)

Comment (rsalevsky) [inline crowbar_framework/lib/crowbar/backup/export.rb:110]: Isn't it enough to keep the permissions of the existing files while restoring?

Comment (vuntz) [inline crowbar_framework/lib/crowbar/backup/export.rb:110]: Files might not exist in the new install :/

**Comment by rhafer** [comment]
> Same here, I'm okay merging if there's a followup PR to address the comments.

@vuntz So, I guess I can consider that the second +1?

**Comment by vuntz** [comment]
> @vuntz So, I guess I can consider that the second +1?

Yes
